src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.cs

```diff
@@ -250,47 +254,62 @@ public void Initialize(Solution solution)
                 // transaction. While some writes can be lost, they are never reordered, and higher layers will recover from that.
                 connection.ExecuteCommand("pragma synchronous=normal", throwOnError: false);
 
-                // First, create all our tables
-                connection.ExecuteCommand(
-$@"create table if not exists ""{StringInfoTableName}"" (
+                CreateTables(connection, MainDBName);
+
+                // Also get the known set of string-to-id mappings we already have in the DB.
+                // Do this in one batch if possible.
+                var fetched = TryFetchStringTableFromMainDB(connection);
+
+                // If we weren't able to retrieve the entire string table in one batch,
+                // attempt to retrieve it for each 
+                var fetchStringTable = !fetched;
+
+                // Try to bulk populate all the IDs we'll need for strings/projects/documents.
+                // Bulk population is much faster than trying to do everything individually.
+                BulkPopulateIds(connection, solution, fetchStringTable);
+
+                // Create and initialize the in-memory write-cache.
+                // From: https://www.sqlite.org/sharedcache.html
+                // Enabling shared-cache for an in-memory database allows two or more database 
+                // connections in the same process to have access to the same in-memory database.
+                // An in-memory database in shared cache is automatically deleted and memory is
+                // reclaimed when the last connection to that database closes.
+
+                connection.ExecuteCommand($"ATTACH DATABASE 'file::memory:?cache=shared' AS ${WriteCacheDBName};");
```

Comment:
@ericsink as far as i can tell from my reading, this is appropriate.  However, one concern i have is that we open the connection with: 

```c#
var flags = OpenFlags.SQLITE_OPEN_CREATE | OpenFlags.SQLITE_OPEN_READWRITE | OpenFlags.SQLITE_OPEN_NOMUTEX;
            var result = (Result)raw.sqlite3_open_v2(databasePath, out var handle, (int)flags, vfs: null);
```

I'm slightly concerned with how nomutex might play with shared_cache.  Is this safe?  Each *connection* will only be used by a single thread at a time.  However, the *in memory* table is *shared* and thus could be written to by many connections simultaneously.  Is this safe? 

Thanks!

Reply:
As I understand it, yes, this is safe. But I wouldn't feel confident in that answer until testing confirms it.

Reply:
I want to describe one thing again that I very briefly mentioned in our conversation on gitter.

I think it might be worth testing your new code in two different configurations.  The architecture is the same either way:  Two SQLite databases attached to one connection.  One serves as a fast write cache.  The other is the real db.

The only difference I suggest exploring is instead of a memory db as the write cache, try using a second SQLite file with PRAGMA synchronous=OFF.  The regular db would stay synchronous=NORMAL.

And the main reason to suggest this is simply because if I were in your shoes, I would be curious about the perf comparison, and it's only a few lines of code different from what you've got.

Note that this idea is not the "temp table" idea, which also got some air time on gitter.  I don't think that path is worth pursuing.

Reply:
> The only difference I suggest exploring is instead of a memory db as the write cache, try using a second SQLite file with PRAGMA synchronous=OFF. The regular db would stay synchronous=NORMAL.

My concern here is now:

1. we have to manage teh lifecycle of the second database.  using `:memory:` or having a temp-db means it always goes away.  Yaay.
2. synchronous=off means we could have undetectable corruption.  That does worry me...